File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:

        current = head
        count = 0
        while current and count < k:
            current = current.next
            count += 1
        if count < k:
            return head

        prev = None
        current = head

        for i in range(k):
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        
        head.next = self.reverseKGroup(current, k)

        return prev

        

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {

        int count = 0;
        ListNode temp = head;

        while(temp!=null && count<k){
            temp = temp.next;
            count++;
        }

        if(count < k) return head;

        ListNode prev = null, next = null;
        ListNode current = head;

        for(int i= 0; i< k; i++){
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }

        ListNode newHead = prev;
        head.next = reverseKGroup(temp, k);
        return newHead;
        
    }
}